data/thermoml_archive/transform.py

```diff
@@ -0,0 +1,140 @@
+import tqdm
+import yaml
+from thermopyl import Parser
+
+from chemnlp.data_val.model import Dataset
+
+
+def get_and_transform_data():
+    """Downloads the archived version of ThermoML, extracts it and
+    parses the provided XML files with thermopyl to construct a flat csv.
+
+    """
+    # get raw data
+    fname = "ThermoML.v2020-09-30.tgz"
+    download_path = pathlib.Path(__file__).parent / fname
+    remote_data_path = f"https://data.nist.gov/od/ds/mds2-2422/{fname}"
+    sha256_checksum = "231161b5e443dc1ae0e5da8429d86a88474cb722016e5b790817bb31c58d7ec2"
+    final_csv_path = pathlib.Path(__file__).parent / "thermoml_archive.csv"
+    final_expected_csv_checksum = ""
+
+    if not download_path.exists():
+        data = requests.get(remote_data_path)
+        with open(download_path, "wb") as f:
+            for chunk in tqdm.tqdm(
+                data.iter_content(chunk_size=8192), desc="Downloading archive"
+            ):
+                f.write(chunk)
+
+    # check if checksum is correct
+    sha256 = hashlib.sha256()
+    with open(download_path, "rb") as f:
+        for chunk in tqdm.tqdm(iter(lambda: f.read(8192), b""), desc="Checking hash"):
+            sha256.update(chunk)
+
+    if received_hash := sha256.hexdigest() != sha256_checksum:
+        raise RuntimeError(
+            "Downloaded file did not match expected checksum -- "
+            "either a new version has been released or something has gone wrong!\n"
+            f"Expected: {sha256_checksum}\n"
+            f"Received: {received_hash}"
+        )
+
+    # Extract tar.gz archive
+    with tarfile.open(download_path, "r:*") as tar:
+        tar.extractall(pathlib.Path(__file__).parent)
+
+    # Loop through journal DOI folders and scrape files
+
+    if final_csv_path.exists():
+        sha256 = hashlib.sha256()
+        with open(final_csv_path, "rb") as f:
+            for chunk in tqdm.tqdm(
+                iter(lambda: f.read(8192), b""), desc="Checking hash"
+            ):
+                sha256.update(chunk)
+        if sha256.hexdigest() != final_expected_csv_checksum:
+            warnings.warn(
+                "Old CSV file did not match expected checksum, will try to recreate."
+            )
+        final_csv_path.rename(final_csv_path.with_suffix(".old.csv"))
+
+    root_dois = ("10.1007", "10.1016", "10.1021")
+
+    num_points = 0
+    num_failed = 0
+    for doi in root_dois:
+        for path in tqdm.tqdm(
+            (pathlib.Path(__file__).parent / doi).glob("*.xml"),
+            desc=f"Looping over files in {doi}",
+        ):
+            with open(path, "r") as f:
+                try:
+                    pd.DataFrame(Parser(path).parse()).to_csv(final_csv_path, mode="a")
+                    num_points += 1
+                except Exception:
+                    num_failed += 1
+
+    print(f"Ingested {num_points} with {num_failed} failures.")
+
+    sha256 = hashlib.sha256()
+    with open(final_csv_path, "rb") as f:
```

Comment:
another copy-paste of the checksum loop, please consider reusable functions, for example:
```python3
def _file_checksum(fn: str):
  with open(...)
    for chunk in ...
      sha.update(chunk)
  return sha.hex()
```